logshifter/input_test.go

```diff
@@ -0,0 +1,90 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+	"sync"
+	"testing"
+)
+
+// TestInputBlocking ensures that Input never blocks, regardless of the queue
+// consumption rate.
+func TestInputBlocking(t *testing.T) {
+	messageCount := int64(10000)
+	messageSize := 1024
+
+	// Use a small queue to help ensure an input backup.
+	queue := make(chan []byte, 1)
+
+	// Set up a fast output sink to help ensure an input backup.
+	go func() {
+		for {
+			<-queue
+		}
+	}()
+
+	// Collect stats.
+	stats := make(map[string]float64)
+	statsCh := make(chan Stat)
+	stop := make(chan struct{})
+	statsWg := sync.WaitGroup{}
+	statsWg.Add(1)
+	go func() {
+		for {
+			select {
+			case <-stop:
+				statsWg.Done()
+				return
+			case s := <-statsCh:
+				stats[s.name] = stats[s.name] + s.value
+			}
+		}
+	}()
+
+	// Read until the test reader closes.
+	reader := newTestReader(messageCount, messageSize)
+	input := &Input{
+		reader:       reader,
+		queue:        queue,
+		bufferSize:   messageSize,
+		statsChannel: statsCh,
+	}
+	input.read()
+
+	// Shut down the stats collector.
+	close(stop)
+	statsWg.Wait()
+	t.Logf("stats: %#v", stats)
```

Comment:
Do I understand correctly that the way this test would fail would be by blocking so that control never reached this point? Might be worth adding a comment that if we get here without deadlock, the test has succeeded.